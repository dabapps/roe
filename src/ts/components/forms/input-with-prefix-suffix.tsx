import * as classNames from 'classnames';
import * as React from 'react';
import { PureComponent } from 'react';
import { ComponentProps } from '../../types';
import InputGroup from './input-group';
import InputGroupAddon from './input-group-addon';

export interface InputWithPrefixSuffixProps extends ComponentProps {
  /**
   * prefix
   */
  prefix?: string;
  /**
   * suffix
   */
  suffix?: string;
  /**
   * input value
   */
  value: any;
  /**
   * is disabled
   */
  disabled?: boolean;
  /**
   * input type
   */
  type?: string;
  /**
   * on change function
   */
  onChange(): void;
}

export class InputWithPrefixSuffix extends PureComponent<InputWithPrefixSuffixProps, {}> {

  public render() {
    const {
      prefix,
      suffix,
      disabled,
      type,
      onChange,
      value,
    } = this.props;

    return (
      <InputGroup>
        {prefix && <InputGroupAddon>{prefix}</InputGroupAddon>}
        <input onChange={onChange} value={value} disabled={disabled} type={type} />
        {suffix && <InputGroupAddon>{suffix}</InputGroupAddon>}
      </InputGroup>
    );
  }
}

export default InputWithPrefixSuffix;
