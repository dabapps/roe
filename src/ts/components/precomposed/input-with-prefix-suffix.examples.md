#### Example

Example with class names

```js
<InputWithPrefixSuffix
  prefix="Â£"
  suffix="%"
  value="Example"
  className="applied-to-input-group"
  inputClassName="applied-to-input"
  prefixClassName="primary"
  suffixClassName="error"
/>
```

Display block with React element prefix

```js
<InputWithPrefixSuffix
  block
  prefix={<strong>Strong</strong>}
  value="Example"
/>
```
