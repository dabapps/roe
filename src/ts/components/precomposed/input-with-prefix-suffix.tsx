import * as React from 'react';

import { OptionalComponentPropAndHTMLAttributes } from '../../types';
import InputGroup from '../forms/input-group';
import InputGroupAddon from '../forms/input-group-addon';

export type InputWithPrefixSuffixProps = {
  /**
   * Content to display to the left of the input.
   */
  prefix?: React.ReactChild;
  /**
   * Content to display to the right of the input.
   */
  suffix?: React.ReactChild;
  /**
   * Set the style `display: block;` so the input group fills its parent.
   */
  block?: boolean;
  /**
   * Class name to apply to the input.
   */
  inputClassName?: string;
  /**
   * Class name to apply to the prefix.
   */
  prefixClassName?: string;
  /**
   * Class name to apply to the suffix.
   */
  suffixClassName?: string;
  /**
   * Input type
   */
  type?: string;
  /**
   * Input value
   */
  value?: string | string[] | number;
  /**
   * Input change handler
   */
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
} & OptionalComponentPropAndHTMLAttributes;

/**
 * A pre-composed Input containing an optional prefix (InputGroupAddon), an input,
 * and an optional suffix (InputGroupAddon).
 */
const InputWithPrefixSuffix = (props: InputWithPrefixSuffixProps) => {
  const {
    prefix,
    suffix,
    block,
    inputClassName,
    prefixClassName,
    suffixClassName,
    component = 'div',
    type,
    value,
    onChange,
    ...remainingProps
  } = props;

  return (
    <InputGroup {...remainingProps} component={component} block={block}>
      {typeof prefix !== 'undefined' && (
        <InputGroupAddon className={prefixClassName}>{prefix}</InputGroupAddon>
      )}
      <input
        className={inputClassName}
        type={type}
        value={value}
        onChange={onChange}
      />
      {typeof suffix !== 'undefined' && (
        <InputGroupAddon className={suffixClassName}>{suffix}</InputGroupAddon>
      )}
    </InputGroup>
  );
};

export default React.memo(InputWithPrefixSuffix);
