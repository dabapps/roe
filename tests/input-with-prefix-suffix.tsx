import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { InputWithPrefixSuffix } from '../src/ts/components/precomposed/input-with-prefix-suffix';

describe('InputWithPrefixSuffix', () => {
  it('should match snapshot', () => {
    const tree = renderer.create(<InputWithPrefixSuffix />);

    expect(tree).toMatchSnapshot();
  });

  it('should apply class names to appropriate elements', () => {
    const tree = renderer.create(
      <InputWithPrefixSuffix
        prefix="£"
        suffix="%"
        className="group"
        inputClassName="input"
        prefixClassName="prefix"
        suffixClassName="suffix"
      />
    );

    expect(tree).toMatchSnapshot();
  });

  it('should add the "block" class name if this prop is true', () => {
    const tree = renderer.create(
      <InputWithPrefixSuffix prefix="£" suffix="%" block />
    );

    expect(tree).toMatchSnapshot();
  });

  it('should pass props to the input', () => {
    const tree = renderer.create(
      <InputWithPrefixSuffix prefix="£" suffix="%" value="Value" />
    );

    expect(tree).toMatchSnapshot();
  });
});
