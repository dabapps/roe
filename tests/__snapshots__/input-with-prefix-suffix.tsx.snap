// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputWithPrefixSuffix should add the "block" class name if this prop is true 1`] = `
<div
  className="input-group block"
>
  <div
    className="input-group-addon"
    style={
      Object {
        "width": undefined,
      }
    }
  >
    £
  </div>
  <input />
  <div
    className="input-group-addon"
    style={
      Object {
        "width": undefined,
      }
    }
  >
    %
  </div>
</div>
`;

exports[`InputWithPrefixSuffix should apply class names to appropriate elements 1`] = `
<div
  className="input-group group"
>
  <div
    className="input-group-addon prefix"
    style={
      Object {
        "width": undefined,
      }
    }
  >
    £
  </div>
  <input
    className="input"
  />
  <div
    className="input-group-addon suffix"
    style={
      Object {
        "width": undefined,
      }
    }
  >
    %
  </div>
</div>
`;

exports[`InputWithPrefixSuffix should match snapshot 1`] = `
<div
  className="input-group"
>
  <input />
</div>
`;

exports[`InputWithPrefixSuffix should pass props to the input 1`] = `
<div
  className="input-group"
>
  <div
    className="input-group-addon"
    style={
      Object {
        "width": undefined,
      }
    }
  >
    £
  </div>
  <input
    onChange={[Function]}
    value="Value"
  />
  <div
    className="input-group-addon"
    style={
      Object {
        "width": undefined,
      }
    }
  >
    %
  </div>
</div>
`;
